ON HOLD: Refactor general excepts

lfdversluis opened the pull request:
A lot of general try excepts are in the wx code which I will not trace because 1. this will be deleted soon 2. most of them are impossible to find qua what they can raise.

Some code chunks are too big or calling C libraries for which no doc is enclosed.

Also note that a lot of general excepts are in the submodules which this PR will not cover.

Contributes, but probably does not fix #1823

Fixes #2183 - **1** bug
Adds **6** tests
Makes sure no bugs under the carpet are swallowed (enhancement)

lfdversluis commented:
retest this please

lfdversluis commented:
Retest this please

lfdversluis commented:
`test_rss.xml` has a internet .png dependency (http://design.ubuntu.com/wp-content/uploads/ubuntu-logo14.png). This one could not be retrieved temporarily...

lfdversluis commented:
retest this please

lfdversluis commented:
retest this please

whirm commented on Tribler/Category/Category.py:
```diff
@@ -89,14 +90,8 @@ def calculateCategory(self, torrent_dict, display_name):
-                return 'xxx'
-        except:
-            self._logger.critical(
-                'Category: Exception in explicit terms filter in torrent: %s', display_name, exc_info=True)
```
Here it would return None if an exception was raised. Have you checked for that on all the callers of this fn?

lfdversluis replied:
I am not sure what you mean by returning None, but the [ function](https://github.com/Tribler/tribler/blob/3ba145c31e79d886b27e6ab4d426217c86ec6e4f/Tribler/Category/FamilyFilter.py#L45) doesn't seem to raise stuff. Only that a[0] is scary, but if that happens we are passing wrong arguments to the function (imo) anyway.

whirm replied:
Oh, brainfart, it's dealt with in 96, sorry for the noise.

whirm commented on Tribler/Category/FamilyFilter.py:
```diff
@@ -33,7 +33,7 @@ def initTerms(self, filename):
             f.close()
-        except:
+        except IOError:
             self._logger.exception(u"Failed to init terms.")
```
Maybe it should say that it couldn't open the file, etc?

whirm commented on Tribler/Core/DownloadConfig.py:
```diff
@@ -191,7 +192,7 @@ def load(filename):
         try:
             dlconfig.read_file(filename)
-        except:
+        except ParsingError:
```
What if the file is not there, do you still get the ParsingError?

whirm commented on Tribler/Core/Utilities/utilities.py:
```diff
@@ -167,7 +167,7 @@ def isValidTorrentFile(metainfo):
         return True
-    except:
+    except ValueError:
         print_exc()
```
This shouldn't print stuff (I know it's not related to this PR, but maybe you want to create a new one fixing this to use the logging system).

whirm commented on Tribler/Core/Video/VideoPlayer.py:
```diff
@@ -159,7 +159,7 @@ def launch_video_player(self, cmd, download=None):
                 self.player_out, self.player_in = os.popen2(cmd, 'b')
-            except:
+            except OSError:
                 print_exc()
```
idem

whirm commented on Tribler/Main/tribler_main.py:
```diff
@@ -879,7 +877,7 @@ def OnExit(self):
                 torrent_db._db.clean_db(randint(0, 24) == 0, exiting=True)
-            except:
+            except (OperationNotEnabledByConfigurationException, ValueError):
                 print_exc()
```
and here

whirm commented:
Looks good!

whirm commented:
retest this please

whirm commented:
@lfdversluis could you rebase it?

lfdversluis commented:
@whirm done.

lfdversluis commented:
@whirm Do you want that print stuff in this PR? Or shall I open an issue and someone can fix it when they have time?

whirm commented:
Its 3 lines, just use the logger with the right level and pass `exc_info=True` to it.

lfdversluis commented:
@whirm Alright, since you sais it's not related to this PR I thought I asked.

whirm commented:
No, it's not, but probably nobody will do it in a loooong time if we create an issue...

lfdversluis commented:
https://jenkins.tribler.org/job/GH_Tribler_PR/939/ ?

lfdversluis commented:
retest this please

whirm commented:
I hope github doesn't ban jenkins.tribler.org D:

whirm commented:
I had to enable repo wiping on all jobs because jenkins' git plugin can't deal with submodule URLs changing so now we are hitting github much harder than before...

lfdversluis commented:
@whirm Ouch, well let's hope more popular projects hit them harder so they won't notice us... 
Also, isn't there an issue tracker for that plugin?

whirm commented:
We are doing shallow clones but each PR run means ~6 clones due to the GHPR tester plugin not getting the variables from the parent job if each job doesn't clone the job itself.

whirm commented:
@lfdversluis yes, I just am too busy to do a proper report on it. And for the other issue, we are using a combo of plugins and I don't even know who's fault it is and I don't have time to investigate it either.

whirm commented:
retest this please

whirm commented on Tribler/Core/Utilities/utilities.py:
```diff
@@ -171,7 +171,7 @@ def isValidTorrentFile(metainfo):
         return True
     except ValueError:
-        print_exc()
+        logger.exception("Could not check torrentfile: a ValueError was thrown", exc_info=True)
```
If you use exception() it's redundant to pass `exc_info`

lfdversluis replied:
Ah you are right. "
Logger.exception(msg, *args, **kwargs)

    Logs a message with level ERROR on this logger. The arguments are interpreted as for debug(), except that any passed exc_info is not inspected. Exception info is always added to the logging message. This method should only be called from an exception handler." 

Removing it now :p

lfdversluis commented:
```
TypeError: expected string or buffer
watchog twisted thread *******TRIPPED!******, hasn't been set for 10.0005 secs.
disabling it and printing traces for all threads.:

*** STACKTRACE - START ***


### ThreadID: 140261498988288 Thread name: torrent_checker
Locals by frame, innermost last:
```

lfdversluis commented:
retest this please

whirm commented:
That comment doesn't help a lot, what was the twisted thread doing when it got stuck?

lfdversluis commented:
https://jenkins.tribler.org/job/GH_Tribler_PR_tests_linux/897/artifact/output/test_runners_output/00006.err/*view*/

```
### ThreadID: 140263005951744 Thread name: Twisted
Locals by frame, innermost last:
/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/dispersy/taskmanager.py:85 is_pending_task_active:
/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/dispersy/taskmanager.py:122 _maybe_clean_task_list:
/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/dispersy/taskmanager.py:57 register_task:
/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Core/APIImplementation/threadpoolmanager.py:35 <lambda>:
/usr/lib/python2.7/dist-packages/twisted/internet/base.py:798 runUntilCurrent:
/usr/lib/python2.7/dist-packages/twisted/internet/base.py:1203 mainLoop:
/usr/lib/python2.7/dist-packages/twisted/internet/base.py:1194 run:
/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Core/Utilities/twisted_thread.py:55 _reactor_runner:
/usr/lib/python2.7/threading.py:754 run:
/usr/lib/python2.7/threading.py:801 __bootstrap_inner:
/usr/lib/python2.7/threading.py:774 __bootstrap:
```
The threadpool was idle. 
Looks like the VideoHTTPServerThread was also doing something

lfdversluis commented:
``test_download_torrent_from_file (Tribler.Test.API.test_download.TestDownload) ... `` again. This confirms (I had seen it elsewhere but couldn't remember where) it's not only on my HTTP tracker async stuff crashing this test.

lfdversluis commented:
retest this please

lfdversluis commented:
`test_download_torrent_from_file (Tribler.Test.API.test_download.TestDownload) ...`

lfdversluis commented:
retest this please

lfdversluis commented:
`test_download_torrent_from_file (Tribler.Test.API.test_download.TestDownload) ... `, `AssertionError: TestMyChannel.test_metadata - No thumbnail torrent in Home panel - Condition was not satisfied in 20 seconds`

lfdversluis commented:
retest this please

lfdversluis commented:
`TestMyChannel.test_metadata - No thumbnail torrent in Home panel - Condition was not satisfied in 20 seconds`

lfdversluis commented:
retest this please

lfdversluis commented:
retest this please

lfdversluis commented:
`Tribler.Test.Core.Upgrade.test_torrent_upgrade_63_64.TestUpgrade63to64.test_upgrade_swift_reseeds_dir_no_dir` on the windows machine. `timeout expired before end of test (5.000000 s.)`

lfdversluis commented:
`File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Core/Modules/tracker_manager.py", line 98, in update_tracker_info
    tracker_info = self._tracker_dict[tracker_url]
exceptions.TypeError: 'NoneType' object has no attribute '__getitem__`

lfdversluis commented:
`Build timed out (after 14 minutes). Marking the build as failed.`

lfdversluis commented:
retest this please

lfdversluis commented:
`TestMyChannel.test_metadata - No thumbnail torrent in Home panel - Condition was not satisfied in 20 seconds`

lfdversluis commented:
retest this please

lfdversluis commented:
```test_anon_download_exitnode_changeofmind (Tribler.Test.test_tunnel_community_negative.TestTunnelCommunityNegative) ... Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Core/Utilities/twisted_thread.py", line 55, in _reactor_runner
    reactor.run(installSignalHandlers=False)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1194, in run
    self.mainLoop()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1203, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 798, in runUntilCurrent
    f(*a, **kw)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/dispersy/endpoint.py", line 250, in dispersythread_data_came_in
    u"standalone_ep")
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/dispersy/dispersy.py", line 1464, in on_incoming_packets
    (str(packet[0]), repr(packet[1])) for packet in packets]
exceptions.AssertionError: [('<Candidate 87.195.128.123:50973>', "'A\\x00\\x04\\x0e\\x00y\\xc6\\xa0\\x00\\x00\\x00\\x00\\x00\\x10\\x00\\x00\\xf3\\xf6\\x00\\x00'")]```

Never seen this one before..

lfdversluis commented:
retest this please

lfdversluis commented:
`TestMyChannel.test_metadata - No thumbnail torrent in Home panel - Condition was not satisfied in 20 seconds` again...

lfdversluis commented:
retest this please

lfdversluis commented:
`TestMyChannel.test_metadata - No thumbnail torrent in Home panel - Condition was not satisfied in 20 seconds` again

lfdversluis commented:
retest this please

lfdversluis commented:
`test_anon_download (Tribler.Test.test_tunnel_community_positive.TestTunnelCommunityPositive) ...` hang

lfdversluis commented:
retest this please

lfdversluis commented:
Several times ``Test raised 1 unhandled Twisted exceptions, last one was: Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/Utilities/twisted_thread.py", line 55, in _reactor_runner
    reactor.run(installSignalHandlers=False)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1194, in run
    self.mainLoop()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1203, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 798, in runUntilCurrent
    f(*a, **kw)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/RemoteTorrentHandler.py", line 472, in _success_callback
    tdef = TorrentDef.load_from_dict(meta_info)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/TorrentDef.py", line 153, in load_from_dict
    return TorrentDef._create(metainfo)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/TorrentDef.py", line 109, in _create
    maketorrent.copy_metainfo_to_input(t.metainfo, t.input)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/APIImplementation/maketorrent.py", line 334, in copy_metainfo_to_input
    outpath = pathlist2filename(file['path'])
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/APIImplementation/maketorrent.py", line 246, in pathlist2filename
    return fullpath.decode('utf-8')
  File "/usr/lib/python2.7/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
exceptions.UnicodeDecodeError: 'utf8' codec can't decode byte 0xbd in position 92: invalid start byte``

Not seen before either, but I did not change any of the files in the trace here...

lfdversluis commented:
retest this please

lfdversluis commented:
`TestMyChannel.test_metadata - No thumbnail torrent in Home panel - Condition was not satisfied in 20 seconds` I am going around in circles here

lfdversluis commented:
retest this please

lfdversluis commented:
retest this please

lfdversluis commented:
`TestMyChannel.test_metadata - No thumbnail torrent in Home panel - Condition was not satisfied in 20 seconds`

lfdversluis commented:
retest this please

lfdversluis commented:
`TestMyChannel.test_metadata - No thumbnail torrent in Home panel - Condition was not satisfied in 20 seconds`

lfdversluis commented:
retest this please

devos50 commented:
retest this please

lfdversluis commented:
`TestMyChannel.test_metadata - No thumbnail torrent in Home panel - Condition was not satisfied in 20 seconds` what a surprise.

lfdversluis commented:
retest this please

lfdversluis commented:
@devos50 can you review this please?

devos50 commented on Tribler/Category/FamilyFilter.py:
```diff
@@ -33,8 +33,8 @@ def initTerms(self, filename):
-        except:
-            self._logger.exception(u"Failed to init terms.")
+        except IOError:
+            self._logger.exception(u"Could not open the file, initTerms failed.")
```
I would add the `filename` to the exception.

devos50 commented on Tribler/Core/DownloadConfig.py:
```diff
@@ -191,7 +192,7 @@ def load(filename):
             dlconfig.read_file(filename)
-        except:
+        except (ParsingError, IOError):
             raise IOError, "Failed to open download config file"
```
Same, it would be nice to add the `filename`.

devos50 commented on Tribler/Core/Utilities/tracker_utils.py:
```diff
@@ -104,7 +104,7 @@ def parse_tracker_url(tracker_url):
             port = int(port)
-        except:
+        except ValueError:
             raise RuntimeError(u'Invalid port number.')
```
I'm not sure whether we should throw a `RuntimeError` here. I think a `ValueError` would be closer the the actual error.

whirm replied:
I would raise a ValueError, we are passing the wrong value to it.

whirm replied:
we can just not catch this exception if we are raising it anyway

lfdversluis replied:
I then vote to remove the try except entirely.

whirm replied:
Go for it

devos50 commented on Tribler/Test/test_my_channel.py:
```diff
@@ -156,7 +156,7 @@ def take_screenshot_quit():
         def check_home():
             self.guiUtility.ShowPage('home')
-            self.CallConditional(20, lambda: len(self.guiUtility.frame.home.aw_panel.list.GetItems()) == 1,
+            self.CallConditional(20, lambda: len(self.guiUtility.frame.home.aw_panel.list.GetItems()) >= 1,
```
Only one item should be added the the `ArtworkPanel` so I think we should leave the equality check.

devos50 commented:
Done reviewing!

lfdversluis commented:
@whirm can you review?

lfdversluis commented:
@whirm I edited the print_exec changes commit, split it into three chunks and did a fixup to those files where you mentioned them. Should be good now for review.

lfdversluis commented:
retest this please

lfdversluis commented:
```
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 798, in runUntilCurrent
    f(*a, **kw)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@3/tribler/Tribler/Main/tribler_main.py", line 464, in define_communities
    kargs=tunnel_kwargs)[0]
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@3/tribler/Tribler/dispersy/dispersy.py", line 407, in define_auto_load
    community = community_cls.init_community(self, master, my_member, *args, **kargs)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@3/tribler/Tribler/dispersy/community.py", line 226, in init_community
    community.initialize(*args, **kargs)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@3/tribler/Tribler/community/tunnel/tunnel_community.py", line 302, in initialize
    self.socks_server.start()
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@3/tribler/Tribler/community/tunnel/Socks5/server.py", line 310, in start
    self.twisted_ports.append(reactor.listenTCP(port, factory))
  File "/usr/lib/python2.7/dist-packages/twisted/internet/posixbase.py", line 478, in listenTCP
    p.startListening()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/tcp.py", line 984, in startListening
    raise CannotListenError(self.interface, self.port, le)
twisted.internet.error.CannotListenError: Couldn't listen on any:30849: [Errno 98] Address already in use
```

lfdversluis commented:
retetst this please

lfdversluis commented:
retest this please

whirm commented:
retest this please

whirm commented:
@lfdversluis maybe the socks test or the server itself has a hardcoded port.

whirm commented:
retest this please

lfdversluis commented:
Build is executing for 3 hours now, killing it

lfdversluis commented:
retest this please

lfdversluis commented:
retest this please

lfdversluis commented:
`Unable to access the X Display, is $DISPLAY set properly?`

lfdversluis commented:
retest this please

lfdversluis commented:
@whirm I just removed the MSVCR100 dependency on the nsi file as @devos50 recompiled the libs. Ready for review.

whirm commented:
Nice!

whirm commented on Tribler/Core/Modules/tracker_manager.py:
```diff
@@ -95,6 +95,11 @@ def update_tracker_info(self, tracker_url, is_successful):
         """
+        # It is possible that a Deferred call could not be cancelled and invokes this function while
+        # we are shutting down, in that case we ignore it.
+        if not self._tracker_dict:
```
I'd rather register the call in the TaskManager so it gets cancelled when shutting down.

lfdversluis replied:
But what would happen if the shutdown already happened and a Deferred fires?

lfdversluis replied:
@whirm ^

whirm replied:
It won't because the delayed call has been canceled with cancel_all_pending_tasks()

lfdversluis replied:
@whirm sometimes a deferred fires despite calling the delay, does the taskmanager handle this?

whirm replied:
If it's canceled it won't fire. If it fires it's too late to cancel it and trying to do so would raise an exception, so the task manager checks before trying. In any case I don't see a Deferred here. What call stack triggered the error? This looks like it should be solved elsewhere to me. 
Quick glance shows me that this function is called from two places. And both of them are running on the reactor's thread. so the decorator is also redundant here.

lfdversluis replied:
I have investigated the torrent_checker where the calls come from thoroughly and everything seems good there. Every time a session is created it is added to the list, and on `shutdown` a loop runs over the list calling the `cleanup` for every session, which cancels the deferred. So it is really weird that it is firing. Also, the torrent checker shutdown happens before the tracker manager in launchmanycores.

I have removed the redundant code btw.

whirm commented:
Almost there! :)

lfdversluis commented:
With the aid of @devos50 the problem has been found. The redundant `@call_on_reactor_thread` decorator above the `update_tracker_info` function was rescheduling the call and sometimes it would be scheduled after the shutdown apparently. Removing this decorator should fix it.

lfdversluis commented:
`exceptions.AssertionError: [('<Candidate 99.122.116.227:61064>', "'A\\x00]\\xd3\\xabi[\\x8d\\x00\\x00\\x00\\x00\\x00\\x10\\x00\\x00\\xb2\\xad\\x00\\x00'")]`

lfdversluis commented:
retest this please

whirm commented:
You rebased again. 

Please point me to the changes that have happened since my last review.

lfdversluis commented:
I only dropped the commit where I added that dict is None check. No other changes were applied as I had already removed that decorator in d8bad50 :)

whirm commented:
Ok then!
Squash and I'll merge

lfdversluis commented:
@whirm Do you want the refactorings of the general execptions on a commit per file base like it is now?

whirm commented:
It's fine, just squash the fixup commits.

lfdversluis commented:
@whirm I am hesitating whether I should split up the last commit into two where the 1) remove stuff goes into its own commit and 2) the additional error in the trap is a fixup for f3367de

or should I target it as a fixup for f3367de entirely / leave it be?

whirm commented:
the catch exception belongs with the others and the decorator thing on its own

lfdversluis commented:
@whirm done!

whirm commented:
This this is failing in several places

lfdversluis commented:
Interesting since I rebased ontop of devel. I am now running git bisect with nose to see where it goes wrong.

lfdversluis commented:
```
git bisect start
git bisect good 03685
git bisect bad 0f83526
git bisect run nosetests Tribler/Test/
e4fb07e2bf1541d29284f56600bec4d5cdcf3738 is the first bad commit
commit e4fb07e2bf1541d29284f56600bec4d5cdcf3738
Author: Elric Milon <whirm@gmx.com>
Date:   Mon Jun 6 13:57:58 2016 +0200

    Move the twisted plugins to the right place

    Why whas it there on the first place?

:040000 040000 d5ea7d622179d277079c5f8fc1892836ecb7a69c 433b2794bbfacf1b4e6d9b7131e544aa1a791baf M	twisted
bisect run success
```

lfdversluis commented:
On the container it goes way back apparently..

whirm commented:
I haven't run any bisect but https://jenkins.tribler.org/job/Test_tribler_devel/1287/ which is commit a517e29ce8cadcca6bf4d2947ed5e93b9686d209 (current HEAD) is just fine.

lfdversluis commented:
The container definitely disagrees with those results :p 
This makes it hard for me to track where it went wrong. It is very weird to begin with, everything passed. I split that commit and squashed (verified there were no changes) and rebased on upstream/devel. Now it's broken 👎

whirm commented:
do a diff with the pre-rebase version (check the reflog)

lfdversluis commented:
No weird changes. Besides the merged code form the PRs (plugin, multichain and upload rate fix) the only thing that differs is the location of maketorrent introduced by my PR.

lfdversluis commented:
https://jenkins.tribler.org/job/GH_Tribler_PR_tests_win32/1217/ 
```
java.io.IOException: remote file operation failed: C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win32\tribler at hudson.remoting.Channel@267a4e3a:BuilderTester-Win32: java.io.IOException: Unable to delete 'C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win32\tribler\.git\objects\pack\pack-68bb1735f583f4cd257ca3ea759690b2112c0e31.idx'. Tried 3 times (of a maximum of 3) waiting 0.1 sec between attempts.
```

lfdversluis commented:
@whirm I think the PR testers of windows broke 👎

devos50 commented:
@lfdversluis should be fixed again. A detached git process kept some files open so I killed it.

devos50 commented:
retest this please

lfdversluis commented:
Rebased again and it... works? Guess I or git messed something up. (But what could go wrong with a git rebase upstream/devel? D:)

whirm commented:
retest this please

lfdversluis commented:
`exceptions.AssertionError: [('<Candidate 37.21.223.117:52658>', "'A\\x00\\xaa\\x05;\\x08o\\xc8\\x00\\x00\\x00\\x00\\x00\\x10\\x00\\x00V\\xac\\x00\\x00'")]` 
`exceptions.AttributeError: 'NoneType' object has no attribute 'magnet_close'`

lfdversluis commented:
retest this please

lfdversluis commented:
`Tribler.Test.Core.test_torrentchecker_session.TestTorrentCheckerSession.test_udpsession_invalid_url` `timeout expired before end of test (5.000000 s.)`

lfdversluis commented:
retest this please

lfdversluis commented:
```
Tribler.Test.Core.Modules.RestApi.test_channels_endpoints.TestChannelsEndpoint.test_get_discovered_channels (from nosetests)

u'Test channel 4' != 'Test channel 0'
```

lfdversluis commented:
retest this please

lfdversluis commented:
This PR seems unstable, we are not sure why. The bug fix in this PR will be a separate PR. All individual commits will be cherry picked and put into a separate PR as well if Devel turns out to be stable. We then can identify which commit introduced the bug as it's most likely a race condition or small-chance exception that is no longer swallowed.

lfdversluis commented:
Closing this one, it has been split into several PRs which got merged
